---
title: "Compute reactivities and investigate coverage"
author: "Kent Riemondy RBI"
date: "5/9/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```

```{r libs}
source("../../R/constants.R")
```

```{r load_dat, eval = F}
files <- dir(file.path(data_dir, "mismatches"),
             pattern = "bedgraph.gz",
             recursive = TRUE,
             full.names = TRUE)

mismatch_bgs <- str_subset(files, "mismatches.bedgraph.gz")

a <- read_bedgraph(files[8])
b <- read_bedgraph(files[16])

bed_intersect_allrows <- function(bed_a, bed_b){
    
    bed_intersect(bed_a, bed_b, suffix = c("", ".ut")) -> intersecting
    bed_intersect(bed_a, bed_b, invert = T)  -> non_intersecting
    filter(intersecting, .overlap != 0) -> overlapping
    
    filter(intersecting, .overlap == 0) %>% 
      select(chrom:value) %>% 
      anti_join(overlapping, by = c("chrom", "start", "end")) %>% 
      unique() -> adjacent
    
    bind_rows(overlapping, non_intersecting) %>% 
      bind_rows(adjacent) -> all_dat
    all_dat
}
```

```{r, eval = F}
dat <- bed_intersect_allrows(a, b) %>% 
  mutate(
    value.ut = ifelse(is.na(value.ut),
                  0,
                  value.ut),
    pDMS = value - value.ut) %>% 
  select(-.overlap)

# drop sites with > 5% mutation rate in untreated samples
dat <- filter(dat, 
              value.ut < 0.05)

# drop sites with pDMS == 0 for now (revisit)
dat <- filter(dat,
              pDMS != 0)

dat_bg <- dat %>% 
  select(chrom:end, pDMS)

kentr::write_gztsv(dat_bg, "dms_ratio.bedgraph", col_names = FALSE)
```


```{r}

mdata <- data_frame(fpath = files,
                    fid = basename(files)) %>% 
  separate(., fid, c("id", "s", "strand", "type"), 
           sep = "_", remove = F) %>% 
  select(-s) %>% 
  mutate(treatment = ifelse(str_detect(id, "DMS"),
                            "DMS",
                            "Untreated"),
         genotype = str_split(id, "-", simplify = T) %>% .[, 2],
         strand = ifelse(strand == "pos",  #first strand library
                         "+",
                         "-"),
         type = str_replace(type, ".bedgraph.gz", ""))

matched_mdata <- mdata %>% 
  filter(!str_detect(type, "depth")) %>% 
  select(fpath, genotype, type, strand, treatment) %>% 
  spread(treatment, fpath) 

matched_mdata_depth <- mdata %>% 
  filter(str_detect(type, "depth")) %>% 
  select(fpath, genotype, type, strand, treatment) %>% 
  spread(treatment, fpath) 

matched_mdata <- left_join(matched_mdata,
                           matched_mdata_depth,
                           by = c("genotype", "strand"), 
                           suffix = c("", "_depth"))
                           
# make list of each DMS versus untreated file path
file_list <- map2(matched_mdata$DMS, 
     matched_mdata$Untreated, ~list(.x, .y))

```


## Find dense coverage regions

Methods from recent paper from Kevin Weeks' lab:

>Data were processed using the ShapeMapper software (Siegfried et al., 2014). Apparent mutation rates were calculated at each genomic position by summing the number of mismatches and deletions and dividing by the number of reads overlapping the position. Sequence insertions and ambiguously aligned deletions were excluded. Mutations spanning multiple adjacent nucleotides were treated as single mutations at the 3′-most position (Siegfried et al., 2014). Nucleotides with apparent mutation rates above 0.02 in any untreated sample were excluded from analysis. In some genomic regions, we observed clusters of elevated mutation rates that appeared to correspond to local self-complementarity artifacts, possibly arising from PCR. These artifacts were identified as regions of at least 10 nucleotides in which three or more of the 10 nucleotides showed mutation rates above 0.03 in the absence of 1M7 treatment, or modified mutation rates above 0.1 in any condition, and were also excluded from analysis. Except where noted elsewhere, SHAPE reactivities were only computed for nucleotides possessing sequencing depths above 1000 in both modified and untreated samples; nucleotides not passing this filter were treated as “no data” and excluded from analysis. Genes were required to have SHAPE data across 80% of the coding sequence for a given condition to be included in gene-specific analyses.

```{r ctrl_data}
res <- bed_intersect(dfs[[1]], dfs[[2]])
dplyr::mutate(res, 
              start = pmax(start.x, start.y),
              end = pmin(end.x, end.y))
```

```{r expt_data}
dfs <- map(unique(matched_mdata$DMS_depth), 
           ~read_bedgraph(.x))

dfs <- map(dfs,
           ~filter(.x, value > 100))

dfs <- map2(dfs, matched_mdata$strand[1:6], 
            ~mutate(.x, strand = .y))

dfs <- map2(dfs, rep(seq(1:3), each = 2),
            ~mutate(.x, expt = .y))

merged_expts <- bind_rows(m_dfs)
```