---
title: "Metaplots"
author: "Kent Riemondy RBI"
date: "7/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(tidyverse)
library(txcompare)
library(valr)
source("../../R/constants.R")
```

Goals:

Make metaplots showing the following data:
1) DMS reactivity 
2) Structure score
3) RNA-editing

```{r}
gtf_fn <- "~/Projects/shared_dbases/annotation/gencode.v26.primary_assembly.annotation.gtf"
gtf <- tidy_gtf(gtf_fn)

alus <- read_bed("../../dbases/hg19_rmsk.bed", n_fields = 6) %>% 
  filter(name == "Alu")

chrom_sizes <- "~/Projects/shared_dbases/genomes/human/chr_appended_grch37/Homo_sapiens.GRCh37.dna.primary_assembly_chrappended.fa.fai"
fa_file <- "~/Projects/shared_dbases/genomes/human/chr_appended_grch37/Homo_sapiens.GRCh37.dna.primary_assembly_chrappended.fa"
chrom_sizes <- read_tsv(chrom_sizes,
                        col_names = c(
                          "chrom",
                          "size"
                        ), 
                        col_types = "ci---")
```


```{r}
dms_tbl_dir <- file.path(data_dir,
                         "mismatches",
                         "july2019", 
                         "filtered_tables")
dms_files <- dir(file.path(dms_tbl_dir, "15"),
                 recursive = TRUE,
                 pattern = ".bgz$",
                 full.names = TRUE)


```

```{r}

zscore <- function(x, na.rm = TRUE){
  (x - mean(x, na.rm = na.rm)) / sd(x, na.rm = na.rm)
}


tbx <- dms_files[1]
a <- read_tabix(tbx, region = "chr1:1000000-2000000")  

region_df <- data.frame(chrom = "chr1", start = 1e6, end = 2e6)
b <- read_tabix(tbx, region_df = region_df)  
all(a == b)
```


## Get PolyA sites

```{r}
# unclear if these are zero or one based. Will assume 1 based for now
pas_db_bed <- read_tsv("../../dbases/polyAdb_v3/human.PAS.txt") %>% 
  mutate(end = Position,
         start = Position - 1) %>% 
    select(chrom = Chromosome,
         start,
         end,
         gene_name = `Gene Symbol`,
         pas_signal = `PAS Signal`,
         strand = Strand) %>% 
  bed_sort()

# get sequences upstream of Poly(A) site
pas_db_bed_seq <- bed_slop(pas_db_bed, 
                       genome = chrom_sizes, 
                       left = 99,
                       strand = TRUE) %>% 
                         get_sequences(., fa_file)

# find position of PAS
pas_db_bed_seq <- mutate(pas_db_bed_seq, 
                         PAS_pos = str_locate_all(seq, "A[TA]TAAA") %>% 
                           map(., ~.x[nrow(.x), 1]) %>% 
                           map_int(., ~ifelse(length(.x) == 0, NA, .x)),
                         PAS_pos = 101 - PAS_pos
                         )


pas_db <- read_tabix(tbx, region_df = pas_db_bed)

# tabix is reporting two ivls per nucleotide, restict to one matching start? or end?
pas_dms <- inner_join(pas_db_bed_seq, 
                      pas_db, 
                      by = c("chrom" = "X1", 
                             "start" = "X2")) %>% 
  unique()

high_confidence_pas_regions <- pas_dms %>% 
  arrange(desc(X5)) %>% 
  filter(!is.na(PAS_pos), 
         !is.na(gene_name),
         gene_name != "na") %>% 
  bed_slop(genome = chrom_sizes, both = 50) %>%
  filter(strand == "-") %>% 
  unique()
```


```{r}
samples <- c(
  "Slow",
  "WT"
)
ss_tbl_dir <- file.path(data_dir, 
                        "enzymatic_data_july2019", 
                        "struct_score", 
                        "norm_coverage")

ss_files <- file.path(ss_tbl_dir,
                      samples,
                      "single_files",
                      paste0(samples, ".tsv.gz"))
 
ss_pas <- read_tabix(ss_files[1],
                     region_df = high_confidence_pas_regions ) 
colnames(ss_pas)[c(1:3, 11)] <- c("chrom", "start", "end", "strand")
ss_pas <- filter(ss_pas, strand == '-')
ss_pas <- unique(ss_pas)


```

```{r}
dms_pas <- read_tabix(dms_files[2],
                     region_df = high_confidence_pas_regions) 
dms_pas <- mutate(dms_pas, 
                  end = X2,
                  start = X2 - 1L) %>% 
  select(chrom = X1,
         start,
         end, 
         strand = X3,
         react = X22)

dms_pas <- filter(dms_pas, strand == '-')
dms_pas <- unique(dms_pas)

```

```{r}
high_confidence_pas_vector <-mutate(high_confidence_pas_regions,
                                    pas_bin = findInterval(PAS_pos, seq(0, 100, 10))) %>% 
  bed_makewindows(., win_size = 1) %>% 
  bed_sort() %>% 
  group_by(pas_bin)
 
split(high_confidence_pas_vector, 
      high_confidence_pas_vector$pas_bin) %>%   map_dfr(., ~bed_map(.x,
        dms_pas,
        out = react, 
        min_overlap = 1) %>% 
 # mutate(out = mean(out, na.rm = TRUE)) %>% 
  group_by(.win_id) %>% 
  summarize(x = mean(out, na.rm = TRUE)),
 .id = "bin") -> a


a$bin <- as.numeric(a$bin) 

a <- filter(a, .win_id <= 100)

ggplot(a, aes(.win_id, bin)) + 
  geom_tile(aes(fill = x * 100)) +
  scale_fill_viridis_c()
```



# make windows for thes PAS sites 
slopped_pas <- pas_dms %>% 
  bed_slop(genome = chrom_sizes, both = 50) %>%
  filter(strand == "+") %>% 
  head(5) %>% 
  bed_makewindows(., win_size = 1) %>% 
  bed_sort()

a <- read_tabix(tbx, 
                region_df = slopped_pas, 
                file_query = "-T") %>% 
  unique()
```














