---
title: "Evofold validation (Filter for A and C)"
author: "Kent Riemondy RBI"
date: "6/26/2019"
output: html_document
---


```{r libs}
source("../../R/constants.R")
source("../../R/viz.R")
library(broom)

fa_fn <- "~/Projects/shared_dbases/genomes/human/chr_appended_grch37/Homo_sapiens.GRCh37.dna.primary_assembly_chrappended.fa"
folding_dir <- "no_slop/"

outdir <- "plots/A_C_only_no_slop/"
dir.create(outdir, recursive = TRUE)

depths <- c("10", "15", "50", "100")

dp_dirs <- file.path(folding_dir, 
          paste0(depths, "_mapfiles"))

names(depths) <- depths


all_files <- map(depths,
                  function(x){
  list(dms = file.path(folding_dir, "RNAstructure", 
                       paste0(x, "_rnastructure.db")),

       no_dms = file.path(folding_dir, "RNAstructure", 
                       paste0(x, "_rnastructure_nodms.db"))
       )
})
```


```{r}
ref_files <- file.path(folding_dir, paste0(depths, "_cutoff.txt"))
refs <- map(ref_files, read_tsv, 
            col_names = c("chrom",
                          "start",
                          "end",
                          "id",
                          "score",
                          "strand",
                          "size",
                          "ss",
                          "confidence",
                          "n_coverage")) %>% 
  map(~get_sequences(.x, 
                     fa_fn))

n_slop <- 0

read_rnastructure <- function(fn){
  lines <- readr::read_lines(fn)
  first_char <- str_sub(lines, 1, 1)
  hdr_idx <- which(first_char == ">")
  ss_idx <- which(first_char %in% c(".", "(", ")"))

  df <- tibble(header = lines[hdr_idx],
               ss = lines[ss_idx])
  
  df <- mutate(df,
               header = str_remove(header, "^>"),
               header = str_trim(header),
               header = str_split(df$header, " ", simplify = F) %>%
                 map_chr(~.x[length(.x)]),
               header = str_remove(header, "^>"))
  df
}

all_dat <- pmap(list(all_files, refs),
     function(fns, ref){
  n_cut <- n_slop + 1
  dat <- read_rnastructure(fns$dms) %>% 
    separate(header, into = c("depth", "id"), sep = "/") %>% 
    mutate(depth = str_remove(depth, "_mapfiles"),
           id = str_remove(id, "_coverage"))
  
  merged_dat <- left_join(dat, ref, by = "id")
  
  slim_dat <- merged_dat %>%
    select(pred = ss.x,
           ref = ss.y,
           size,
           score,
           id,
           seq) %>%
    mutate(pred = str_sub(pred, n_cut,-n_cut),
           seq = str_sub(seq, n_cut, -n_cut))
  
  tmp <- split(slim_dat, 1:nrow(slim_dat)) %>% map_dfr(~calc_ss_similarity(.x$pred, 
                                  .x$ref, 
                                  .x$seq, 
                                  count_only_ac = TRUE))
  dms_dat <- bind_cols(slim_dat, tmp)
  
  dat <- read_rnastructure(fns$no_dms) %>% 
    separate(header, into = c("depth", "tmp", "id"), sep = "/") %>% 
    mutate(depth = str_remove(depth, "_mapfiles"),
           id = str_remove(id, "_coverage")) %>% 
    select(-tmp)
    
  merged_dat <- left_join(dat, ref, by = "id")
  
  slim_dat <- merged_dat %>%
    select(pred = ss.x,
           ref = ss.y,
           size,
           score, 
           id,
           seq) %>%
    mutate(pred = str_sub(pred, n_cut,-n_cut),
           seq = str_sub(seq, n_cut, -n_cut))
  
  tmp <- split(slim_dat, 1:nrow(slim_dat)) %>% map_dfr(~calc_ss_similarity(.x$pred, 
                                  .x$ref, 
                                  .x$seq, 
                                  count_only_ac = TRUE))
  no_dms_dat <- bind_cols(slim_dat, tmp)
  
  bind_rows(list(
    dms = dms_dat,
    no_dms = no_dms_dat
  ), .id = "treatment")
})

all_dat <- bind_rows(all_dat, .id = "depth")
       


```

```{r size_dist}
p_size <- ggplot(all_dat %>% filter(treatment == "no_dms"), aes(size) ) +
  geom_histogram(bins = 50) + 
  labs(x = "Evofold region size")

p_score <-  ggplot(all_dat %>% filter(treatment == "no_dms"), aes(score) ) +
  geom_histogram(bins = 50) + 
  labs(x = "Evofold region score")

p <- plot_grid(p_size, p_score, nrow = 1, ncol = 2)
save_plot(
    file.path(outdir, "evofold_distribution.pdf"),
    p,
    nrow = 1,
    ncol = 2,
    base_asp = 0.75
  )
p
```

```{r, all_no_filter}


plt_dat <- all_dat

plt_dat <- mutate(plt_dat, 
                  depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))



plot_sen_ppv <-
  function(plt_dat, out_name, outdir, facet_by = NULL,
           cut_var = NULL) {
    tsize = 2
    if (is.null(facet_by)) {
      stat_df <- plt_dat %>%
        nest(-depth) %>%
        mutate(
          res = map(data,
                    ~ wilcox.test(sen ~ treatment, data = .)),
          res = map(res, tidy),
          n = map_int(data, nrow)
        ) %>%
        unnest(res) %>%
        mutate(
          sen = max(plt_dat$sen) * 1.1,
          p.value = paste0("p = ", signif(p.value, 2)),
          group = row_number(),
          xend = group - 0.2,
          xstart = group + 0.2,
          y = max(plt_dat$sen) * 1.025
        )
      
      p_sen <- ggplot(plt_dat,
                      aes(depth, sen)) +
        geom_violin(aes(fill = treatment)) +
        geom_boxplot(aes(fill = treatment),
                     alpha = 0.75,
                     position = position_dodge(0.9)) +
        scale_fill_brewer(palette = "Set1",
                          name = "") +
        geom_text(data = stat_df,
                  aes(label = p.value),
                  size = tsize) +
        geom_segment(data = stat_df, aes(
          x = xstart,
          xend = xend,
          y = y,
          yend = y
        )) +
        labs(x = "Nucleotide Depth",
             y = "Sensitivity") +
        theme_cowplot() +
        theme(legend.position = "top")
      
      stat_df <- plt_dat %>%
        nest(-depth) %>%
        mutate(
          res = map(data,
                    ~ wilcox.test(ppv ~ treatment, data = .)),
          res = map(res, tidy),
          n = map_int(data, nrow)
        ) %>%
        unnest(res) %>%
        mutate(
          ppv = max(plt_dat$ppv) * 1.1,
          p.value = paste0("p = ", signif(p.value, 2)),
          group = row_number(),
          xend = group - 0.2,
          xstart = group + 0.2,
          y = max(plt_dat$ppv) * 1.025
        )
      
      p_ppv <- ggplot(plt_dat,
                      aes(depth, ppv)) +
        geom_violin(aes(fill = treatment)) +
        geom_boxplot(aes(fill = treatment),
                     alpha = 0.75,
                     position = position_dodge(0.9)) +
        geom_text(data = stat_df,
                  aes(label = p.value),
                  size = tsize) +
        geom_segment(data = stat_df, aes(
          x = xstart,
          xend = xend,
          y = y,
          yend = y
        )) +
        scale_fill_brewer(palette = "Set1",
                          name = "") +
        labs(x = "Nucleotide Depth",
             y = "P.P.V") +
        theme_cowplot() +
        theme(legend.position = "top")
    } else {
      # stats
      stat_df <- plt_dat %>%
        nest(-depth,-level) %>%
        mutate(
          res = map(data,
                    ~ wilcox.test(sen ~ treatment, data = .)),
          res = map(res, tidy),
          n = map_int(data, nrow)
        ) %>%
        unnest(res) %>%
        arrange(level, depth) %>%
        group_by(level) %>%
        mutate(
          sen = max(plt_dat$sen) * 1.1,
          p.value = paste0("p = ", signif(p.value, 2)),
          group = row_number(),
          xend = group - 0.2,
          xstart = group + 0.2,
          y = max(plt_dat$sen) * 1.025
        )
      
      # facet label
      facet_by <- "level"
      rename_facet <- group_by(plt_dat, level) %>%
        summarize(n = n(),
                  x = min(!!sym(cut_var)),
                  y = max(!!sym(cut_var))) %>%
        mutate(new_name = paste0(x, "-", y, " (", n, " regions)")) %>%
        pull(new_name)
      names(rename_facet) <- sort(unique(plt_dat$level))
      
      p_sen <- ggplot(plt_dat,
                      aes(depth, sen)) +
        geom_violin(aes(fill = treatment)) +
        geom_boxplot(aes(fill = treatment),
                     alpha = 0.75,
                     position = position_dodge(0.9)) +
        geom_text(data = stat_df,
                  aes(label = p.value),
                  size = tsize) +
        geom_segment(data = stat_df, aes(
          x = xstart,
          xend = xend,
          y = y,
          yend = y
        )) +
        facet_wrap(as.formula(paste("~ ", facet_by)),
                   labeller = as_labeller(rename_facet)) +
        scale_fill_brewer(palette = "Set1",
                          name = "") +
        labs(x = "Nucleotide Depth",
             y = "Sensitivity") +
        theme_cowplot() +
        theme(legend.position = "top")
      
      
      # stats
      stat_df <- plt_dat %>%
        nest(-depth,-level) %>%
        mutate(
          res = map(data,
                    ~ wilcox.test(ppv ~ treatment, data = .)),
          res = map(res, tidy),
          n = map_int(data, nrow)
        ) %>%
        unnest(res) %>%
        arrange(level, depth) %>%
        group_by(level) %>%
        mutate(
          ppv = max(plt_dat$ppv) * 1.1,
          p.value = paste0("p = ", signif(p.value, 2)),
          group = row_number(),
          xend = group - 0.2,
          xstart = group + 0.2,
          y = max(plt_dat$ppv) * 1.025
        )
      
      # # facet label
      # facet_by <- "level"
      # rename_facet <- group_by(plt_dat, level) %>%
      #   summarize(n = n(),
      #             x = min(!!sym(cut_var)),
      #             y = max(!!sym(cut_var))) %>%
      #   mutate(new_name = paste0(x, "-", y, " ( ", n, "regions)")) %>%
      #   pull(new_name)
      # names(rename_facet) <- unique(plt_dat$level)
      
      
      p_ppv <- ggplot(plt_dat,
                      aes(depth, ppv)) +
        geom_violin(aes(fill = treatment)) +
        geom_boxplot(aes(fill = treatment),
                     alpha = 0.75,
                     position = position_dodge(0.9)) +
        geom_text(data = stat_df, aes(label = p.value),
                  size = tsize) +
        geom_segment(data = stat_df, aes(
          x = xstart,
          xend = xend,
          y = y,
          yend = y
        )) +
        facet_wrap(as.formula(paste("~ ", facet_by)),
                   labeller = as_labeller(rename_facet)) +
        scale_fill_brewer(palette = "Set1",
                          name = "") +
        labs(x = "Nucleotide Depth",
             y = "P.P.V") +
        theme_cowplot() +
        theme(legend.position = "top")
    }
    p <- plot_grid(p_sen, p_ppv, nrow = 1, ncol = 2)
    p
    
    
    save_plot(
      file.path(outdir, out_name),
      p,
      nrow = 1,
      ncol = 2,
      base_asp = 1.2
    )
    p
  }

plot_sen_ppv(plt_dat, "s_ppv_no_filtering.pdf", outdir)

```


```{r, all_no_filter}

not_perfectly_predicted <- all_dat %>%
  filter(treatment == "no_dms") %>% 
  filter(sen < 0.75 | ppv < 0.75) %>% 
  pull(id) %>% 
  unique()

plt_dat <- filter(all_dat, id %in% not_perfectly_predicted)

plt_dat <- mutate(plt_dat, 
                  depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))

plot_sen_ppv(plt_dat, "s_ppv_filtering.pdf", outdir)

```

```{r, cutbyscore}

not_perfectly_predicted <- all_dat %>%
  filter(treatment == "no_dms") %>% 
  filter(sen < 0.75 | ppv < 0.75) %>% 
  pull(id) %>% 
  unique()

plt_dat <- filter(all_dat, id %in% not_perfectly_predicted)

plt_dat <- group_by(plt_dat,
                    depth) %>% 
   mutate(level = cut(score, 
                     quantile(score), 
                     include.lowest = TRUE, 
                     right = FALSE, 
                     labels = FALSE)) %>% 
  ungroup() %>% 
  mutate(depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))

plot_sen_ppv(plt_dat, "s_ppv_filtering_cut_score.pdf", 
             outdir, 
             facet_by = "level", 
             cut_var = "score")

```


```{r}
not_perfectly_predicted <- all_dat %>%
  filter(treatment == "no_dms") %>% 
  filter(sen < 0.75 | ppv < 0.75) %>% 
  pull(id) %>% 
  unique()

plt_dat <- filter(all_dat, id %in% not_perfectly_predicted)

plt_dat <- group_by(plt_dat,
                    depth) %>% 
  mutate(level = cut(size, 
                     quantile(size), 
                     include.lowest = TRUE, 
                     right = FALSE, 
                     labels = FALSE)) %>% 
  ungroup() %>% 
  mutate(depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))


plot_sen_ppv(plt_dat, "s_ppv_filtering_cut_size.pdf", outdir, facet_by = "level", cut_var = "size")

```


## RNAfold


```{r}
outdir <- "plots/A_C_only_no_slop/rnafold"
dir.create(outdir, showWarnings = FALSE)
depths <- c("10", "15", "50", "100")

dp_dirs <- file.path(folding_dir, 
          paste0(depths, "_mapfiles"))

names(depths) <- depths


all_files <- map(depths,
                  function(x){
  list(dms = file.path(folding_dir, "RNAfold", 
                       paste0(x, "_rnafold_wangetal.fasta")),

       no_dms = file.path(folding_dir, "RNAfold", 
                       paste0(x, "_rnafold_nodms.fasta"))
       )
})
```


```{r}
ref_files <- file.path(folding_dir, paste0(depths, "_cutoff.txt"))
refs <- map(ref_files, read_tsv, 
            col_names = c("chrom",
                          "start",
                          "end",
                          "id",
                          "score",
                          "strand",
                          "size",
                          "ss",
                          "confidence",
                          "n_coverage")) %>% 
  map(~get_sequences(.x, 
                     fa_fn))

n_slop <- 0

read_rnafold <- function(fn) {
  x <- read_lines(fn)
 
  first_char <- str_sub(x, 1, 1)
  hdr_idx <- seq(1, length(x), by = 3)
  seq_idx <- seq(2, length(x), by = 3)
  ss_idx <- seq(3, length(x), by = 3)
  
  res <- tibble(header = x[hdr_idx],
         seq = x[seq_idx],
         ss = x[ss_idx])
  res <- mutate(res, 
                ss = str_split(ss, " ", simplify = TRUE) %>% .[, 1],
                header = str_remove(header, "^>"))
  res
}


all_dat <- pmap(list(all_files, refs),
     function(fns, ref){
  n_cut <- n_slop + 1
  dat <- read_rnafold(fns$dms) %>% 
    separate(header, into = c("depth", "id"), sep = "/") %>% 
    mutate(depth = str_remove(depth, "_mapfiles"),
           id = str_remove(id, "_coverage")) %>% 
    select(id, ss, -seq)
  
  merged_dat <- left_join(dat, ref, by = c("id"))
  
  slim_dat <- merged_dat %>%
    select(pred = ss.x,
           ref = ss.y,
           size,
           score,
           id,
           seq) %>%
    mutate(pred = str_sub(pred, n_cut,-n_cut),
           seq = str_sub(seq, n_cut, -n_cut))
  
  tmp <- split(slim_dat, 1:nrow(slim_dat)) %>% map_dfr(~calc_ss_similarity(.x$pred, 
                                  .x$ref, 
                                  .x$seq, 
                                  count_only_ac = TRUE))
  dms_dat <- bind_cols(slim_dat, tmp)
  
  dat <- read_rnafold(fns$no_dms) %>% 
    select(id = header, ss, -seq)
    
  merged_dat <- left_join(dat, ref, by = "id")
  
  slim_dat <- merged_dat %>%
    select(pred = ss.x,
           ref = ss.y,
           size,
           score, 
           id,
           seq) %>%
    mutate(pred = str_sub(pred, n_cut,-n_cut),
           seq = str_sub(seq, n_cut, -n_cut))
  
  tmp <- split(slim_dat, 1:nrow(slim_dat)) %>% map_dfr(~calc_ss_similarity(.x$pred, 
                                  .x$ref, 
                                  .x$seq, 
                                  count_only_ac = TRUE))
  no_dms_dat <- bind_cols(slim_dat, tmp)
  
  bind_rows(list(
    dms = dms_dat,
    no_dms = no_dms_dat
  ), .id = "treatment")
})

all_dat <- bind_rows(all_dat, .id = "depth")
       

plt_dat <- all_dat

plt_dat <- mutate(plt_dat, 
                  depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))



plot_sen_ppv(plt_dat, "s_ppv_no_filtering.pdf", outdir)

```


```{r, all_no_filter}

not_perfectly_predicted <- all_dat %>%
  filter(treatment == "no_dms") %>% 
  filter(sen < 0.75 | ppv < 0.75) %>% 
  pull(id) %>% 
  unique()

plt_dat <- filter(all_dat, id %in% not_perfectly_predicted)

plt_dat <- mutate(plt_dat, 
                  depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))

plot_sen_ppv(plt_dat, "s_ppv_filtering.pdf", outdir)

```
```





## Find regions with better folding with DMS data


```{r}
not_perfectly_predicted <- all_dat %>%
  filter(treatment == "no_dms") %>% 
  filter(sen < 0.75 | ppv < 0.75) %>% 
  pull(id) %>% 
  unique()

plt_dat <- filter(all_dat, id %in% not_perfectly_predicted)

d_dat <- plt_dat %>% filter(depth == 100, size > 20) %>% 
  select(-depth) %>% 
  filter(treatment == "dms")

ut_dat <- plt_dat %>% filter(depth == 100, size > 20) %>% 
  select(-depth) %>% 
  filter(treatment != "dms")

dat <- left_join(d_dat, ut_dat, 
                 by = c("size",
                        "score",
                        "id",
                        "seq",
                        "ref"),
                 suffix = c("_dms", "_ut"))

dat <- mutate(dat, 
              delta_ppv = ppv_dms - ppv_ut,
              delta_sen = sen_dms - sen_ut)

outdir <- "examples"
dir.create(outdir)
plot_pred <- function(df){
  map_dat <- read_map(file.path(folding_dir, "100_mapfiles", 
                                paste0(df$id, "_coverage.map")))
  
  plot_secondary_structure(df$seq, 
                           df$ref,
                           color_by = map_dat$react,
                           outpath = file.path(outdir,
                                               paste0(df$id, 
                                                      "_evofold.png"))
                           )
  
  plot_secondary_structure(df$seq, 
                           df$pred_dms,
                           color_by = map_dat$react,
                           outpath = file.path(outdir,
                                               paste0(df$id, 
                                                      "_with_dms.png"))
                           )
  
    plot_secondary_structure(df$seq, 
                           df$pred_ut,
                           color_by = NULL,
                           outpath = file.path(outdir,
                                               paste0(df$id, 
                                                      "_without_dms.png"))
                           )
}


to_plot <- c(
  "1022231_0_+_95", # CLTC CDS
  "1495619_0_+_59", # HMGXB4 3UTR
  "491952.0_0_-_92", # FBXL14 CDS
  "1932934_0_-_135" #FBN2 CDS)
)
plt_dat <- filter(dat, id %in% to_plot)

walk(split(plt_dat, 1:nrow(plt_dat)), plot_pred)

filter(refs[[3]], id %in% to_plot)
```


