---
title: "DMS validation"
author: "Kent Riemondy RBI"
date: "`R Sys.Date()`"
output: html_document
---


```{r libs}
source("../../R/constants.R")
library(broom)
```

```{r paths}
fa_fn <- "~/Projects/shared_dbases/genomes/human/chr_appended_grch37/Homo_sapiens.GRCh37.dna.primary_assembly_chrappended.fa"
folding_dir <- "all_new_cutoffs"

outdir <- "plots/rnastructure/all_nucs"
dir.create(outdir, recursive = TRUE)

depths <- c("15", "50", "100", "500")

dp_dirs <- file.path(folding_dir, 
          paste0(depths, "_mapfiles"))

names(depths) <- depths
```

## RNAstructure

```{r get_files}
all_files <- map(depths,
                  function(x){
  list(dms = file.path(folding_dir, "RNAstructure", 
                       paste0(x, "_rnastructure.db")),

       no_dms = file.path(folding_dir, "RNAstructure", 
                       paste0(x, "_rnastructure_nodms.db"))
       )
})
```


```{r}
ref_files <- file.path(folding_dir, paste0(depths, "_cutoff.txt"))
refs <- map(ref_files, read_tsv, 
            col_names = c("chrom",
                          "start",
                          "end",
                          "id",
                          "score",
                          "strand",
                          "size",
                          "ss",
                          "confidence",
                          "n_coverage")) %>% 
  map(~get_sequences(.x, 
                     fa_fn))

n_slop <- 0

all_dat <- pmap(list(all_files, refs),
     function(fns, ref){
  # read in strucutre predicitions and merge with reference bedfile with
  # "known" structures
  n_cut <- n_slop + 1
  dat <- read_rnastructure(fns$dms) %>% 
    separate(header, into = c("depth", "id"), sep = "/") %>% 
    mutate(depth = str_remove(depth, "_mapfiles"),
           id = str_remove(id, "_coverage"))
  
  merged_dat <- left_join(dat, ref, by = "id")
  
  slim_dat <- merged_dat %>%
    select(pred = ss.x,
           ref = ss.y,
           size,
           score,
           id,
           seq) %>%
    mutate(pred = str_sub(pred, n_cut,-n_cut),
           seq = str_sub(seq, n_cut, -n_cut))
  
  # calculate sen and ppv 
  tmp <- split(slim_dat, 1:nrow(slim_dat)) %>% 
    map_dfr(~calc_ss_similarity(.x$pred, 
                                  .x$ref, 
                                  .x$seq, 
                                  count_only_ac = FALSE))
  dms_dat <- bind_cols(slim_dat, tmp)
  
  # repeat for predicitions without dms data
  dat <- read_rnastructure(fns$no_dms) %>% 
    separate(header, into = c("depth", "id"), sep = "/") %>% 
    mutate(depth = str_remove(depth, "_mapfiles"),
           id = str_remove(id, "_coverage"))
    
  merged_dat <- left_join(dat, ref, by = "id")
  
  slim_dat <- merged_dat %>%
    select(pred = ss.x,
           ref = ss.y,
           size,
           score, 
           id,
           seq) %>%
    mutate(pred = str_sub(pred, n_cut,-n_cut),
           seq = str_sub(seq, n_cut, -n_cut))
  
  tmp <- split(slim_dat, 1:nrow(slim_dat)) %>%
    map_dfr(~calc_ss_similarity(.x$pred, 
                                  .x$ref, 
                                  .x$seq, 
                                  count_only_ac = FALSE))
  no_dms_dat <- bind_cols(slim_dat, tmp)
  
  bind_rows(list(
    dms = dms_dat,
    no_dms = no_dms_dat
  ), .id = "treatment")
})

all_dat <- bind_rows(all_dat, .id = "depth")
       
```

```{r size_dist}
p_size <- ggplot(all_dat %>% filter(treatment == "no_dms"), aes(size) ) +
  geom_histogram(bins = 50) + 
  labs(x = "Evofold region size")

p_score <-  ggplot(all_dat %>% filter(treatment == "no_dms"), aes(score) ) +
  geom_histogram(bins = 50) + 
  labs(x = "Evofold region score")

p <- plot_grid(p_size, p_score, nrow = 1, ncol = 2)
save_plot(
    file.path(outdir, "evofold_distribution.pdf"),
    p,
    nrow = 1,
    ncol = 2,
    base_asp = 0.75
  )
p
```

```{r}

plot_sen_ppv <- function(plt_dat,
           out_name,
           outdir, 
           facet_by = NULL,
           cut_var = NULL) {
    tsize = 2
    if (is.null(facet_by)) {
      stat_df <- plt_dat %>%
        nest(data = c(treatment, 
                      pred, ref, 
                      size, score,
                      id, seq, sen,
                      ppv)) %>%
        mutate(
          res = map(data,
                    ~ wilcox.test(sen ~ treatment, data = .)),
          res = map(res, tidy),
          n = map_int(data, function(x) {length(unique(x$id))})
        ) %>%
        unnest(res) %>%
        mutate(
          sen = max(plt_dat$sen) * 1.1,
          p.value = paste0("n = ", n, "\n",
                           "p = ", signif(p.value, 2)),
          group = row_number(),
          xend = group - 0.2,
          xstart = group + 0.2,
          y = max(plt_dat$sen) * 1.025
        )
      
      p_sen <- ggplot(plt_dat,
                      aes(depth, sen)) +
        geom_violin(aes(fill = treatment)) +
        geom_boxplot(aes(fill = treatment),
                     alpha = 0.75,
                     position = position_dodge(0.9)) +
        scale_fill_brewer(palette = "Set1",
                          name = "") +
        geom_text(data = stat_df,
                  aes(label = p.value),
                  size = tsize) +
        geom_segment(data = stat_df, aes(
          x = xstart,
          xend = xend,
          y = y,
          yend = y
        )) +
        labs(x = "Nucleotide Depth",
             y = "Sensitivity") +
        theme_cowplot() +
        theme(legend.position = "top")
      
      stat_df <- plt_dat %>%
        nest(data = c(treatment, pred, 
                      ref, size, score, 
                      id, seq, sen, ppv)) %>%
        mutate(
          res = map(data,
                    ~ wilcox.test(ppv ~ treatment, data = .)),
          res = map(res, tidy),
           n = map_int(data, function(x) {length(unique(x$id))})
        ) %>%
        unnest(res) %>%
        mutate(
          ppv = max(plt_dat$ppv) * 1.1,
          p.value = paste0("n = ", n, "\n",
                           "p = ", signif(p.value, 2)),
          group = row_number(),
          xend = group - 0.2,
          xstart = group + 0.2,
          y = max(plt_dat$ppv) * 1.025
        )
      
      p_ppv <- ggplot(plt_dat,
                      aes(depth, ppv)) +
        geom_violin(aes(fill = treatment)) +
        geom_boxplot(aes(fill = treatment),
                     alpha = 0.75,
                     position = position_dodge(0.9)) +
        geom_text(data = stat_df,
                  aes(label = p.value),
                  size = tsize) +
        geom_segment(data = stat_df, aes(
          x = xstart,
          xend = xend,
          y = y,
          yend = y
        )) +
        scale_fill_brewer(palette = "Set1",
                          name = "") +
        labs(x = "Nucleotide Depth",
             y = "P.P.V") +
        theme_cowplot() +
        theme(legend.position = "top")
    } else {
      # stats
      stat_df <- plt_dat %>%
        nest(c(treatment, pred, 
               ref, size, score,
               id, seq, sen, ppv)) %>%
        mutate(
          res = map(data,
                    ~ wilcox.test(sen ~ treatment, data = .)),
          res = map(res, tidy),
          n = map_int(data, nrow)
        ) %>%
        unnest(res) %>%
        arrange(level, depth) %>%
        group_by(level) %>%
        mutate(
          sen = max(plt_dat$sen) * 1.1,
          p.value = paste0("p = ", signif(p.value, 2)),
          group = row_number(),
          xend = group - 0.2,
          xstart = group + 0.2,
          y = max(plt_dat$sen) * 1.025
        )
      
      # facet label
      facet_by <- "level"
      rename_facet <- group_by(plt_dat, level) %>%
        summarize(n = n(),
                  x = min(!!sym(cut_var)),
                  y = max(!!sym(cut_var))) %>%
        mutate(new_name = paste0(x, "-", y, " (", n, " regions)")) %>%
        pull(new_name)
      names(rename_facet) <- sort(unique(plt_dat$level))
      
      p_sen <- ggplot(plt_dat,
                      aes(depth, sen)) +
        geom_violin(aes(fill = treatment)) +
        geom_boxplot(aes(fill = treatment),
                     alpha = 0.75,
                     position = position_dodge(0.9)) +
        geom_text(data = stat_df,
                  aes(label = p.value),
                  size = tsize) +
        geom_segment(data = stat_df, aes(
          x = xstart,
          xend = xend,
          y = y,
          yend = y
        )) +
        facet_wrap(as.formula(paste("~ ", facet_by)),
                   labeller = as_labeller(rename_facet)) +
        scale_fill_brewer(palette = "Set1",
                          name = "") +
        labs(x = "Nucleotide Depth",
             y = "Sensitivity") +
        theme_cowplot() +
        theme(legend.position = "top")
      
      
      # stats
      stat_df <- plt_dat %>%
        nest(data = c(treatment, pred,
               ref, size, score, 
               id, seq, sen, ppv)) %>%
        mutate(
          res = map(data,
                    ~ wilcox.test(ppv ~ treatment, data = .)),
          res = map(res, tidy),
          n = map_int(data, nrow)
        ) %>%
        unnest(res) %>%
        arrange(level, depth) %>%
        group_by(level) %>%
        mutate(
          ppv = max(plt_dat$ppv) * 1.1,
          p.value = paste0("p = ", signif(p.value, 2)),
          group = row_number(),
          xend = group - 0.2,
          xstart = group + 0.2,
          y = max(plt_dat$ppv) * 1.025
        )
      
      # # facet label
      # facet_by <- "level"
      # rename_facet <- group_by(plt_dat, level) %>%
      #   summarize(n = n(),
      #             x = min(!!sym(cut_var)),
      #             y = max(!!sym(cut_var))) %>%
      #   mutate(new_name = paste0(x, "-", y, " ( ", n, "regions)")) %>%
      #   pull(new_name)
      # names(rename_facet) <- unique(plt_dat$level)
      
      
      p_ppv <- ggplot(plt_dat,
                      aes(depth, ppv)) +
        geom_violin(aes(fill = treatment)) +
        geom_boxplot(aes(fill = treatment),
                     alpha = 0.75,
                     position = position_dodge(0.9)) +
        geom_text(data = stat_df, aes(label = p.value),
                  size = tsize) +
        geom_segment(data = stat_df, aes(
          x = xstart,
          xend = xend,
          y = y,
          yend = y
        )) +
        facet_wrap(as.formula(paste("~ ", facet_by)),
                   labeller = as_labeller(rename_facet)) +
        scale_fill_brewer(palette = "Set1",
                          name = "") +
        labs(x = "Nucleotide Depth",
             y = "P.P.V") +
        theme_cowplot() +
        theme(legend.position = "top")
    }
    p <- plot_grid(p_sen, p_ppv, nrow = 1, ncol = 2)
    p
    
    save_plot(
      file.path(outdir, out_name),
      p,
      nrow = 1,
      ncol = 2,
      base_asp = 1.2
    )
    p
}

plt_dat <- all_dat

# just plot 15 and 50 for simplicity

plt_dat <- filter(plt_dat, depth %in% c(15, 50))
  
plt_dat <- mutate(plt_dat, 
                  depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))


plot_sen_ppv(plt_dat, "s_ppv_no_filtering.pdf", outdir)

```


```{r}

not_perfectly_predicted <- all_dat %>%
  filter(treatment == "no_dms") %>% 
  filter(sen < 0.9 | ppv < 0.9) %>% 
  pull(id) %>% 
  unique()

plt_dat <- filter(all_dat, id %in% not_perfectly_predicted)

plt_dat <- filter(plt_dat, depth %in% c(15, 50))

plt_dat <- mutate(plt_dat, 
                  depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))

plot_sen_ppv(plt_dat, "s_ppv_filtering.pdf", outdir)

```

```{r, cutbyscore}

not_perfectly_predicted <- all_dat %>%
  filter(treatment == "no_dms") %>% 
  filter(sen < 0.9 | ppv < 0.9) %>% 
  pull(id) %>% 
  unique()

plt_dat <- filter(all_dat, id %in% not_perfectly_predicted)
plt_dat <- filter(plt_dat, depth %in% c(15, 50))


plt_dat <- group_by(plt_dat,
                    depth) %>% 
   mutate(level = cut(score, 
                     quantile(score), 
                     include.lowest = TRUE, 
                     right = FALSE, 
                     labels = FALSE)) %>% 
  ungroup() %>% 
  mutate(depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))

plot_sen_ppv(plt_dat, "s_ppv_filtering_cut_score.pdf", 
             outdir, 
             facet_by = "level", 
             cut_var = "score")

```


```{r}
not_perfectly_predicted <- all_dat %>%
  filter(treatment == "no_dms") %>% 
  filter(sen < 0.75 | ppv < 0.75) %>% 
  pull(id) %>% 
  unique()

plt_dat <- filter(all_dat, id %in% not_perfectly_predicted)
plt_dat <- filter(plt_dat, depth %in% c(15, 50))

plt_dat <- group_by(plt_dat,
                    depth) %>% 
  mutate(level = cut(size, 
                     quantile(size), 
                     include.lowest = TRUE, 
                     right = FALSE, 
                     labels = FALSE)) %>% 
  ungroup() %>% 
  mutate(depth = factor(depth, 
                        levels = sort(as.numeric(unique(depth)))),
         treatment = ifelse(treatment == "dms",
                            "DMS",
                            "Untreated") %>% 
           factor(., levels = c("Untreated", "DMS")))


plot_sen_ppv(plt_dat, "s_ppv_filtering_cut_size.pdf", outdir, facet_by = "level", cut_var = "size")

```

## Find regions with better folding with DMS data


```{r,}
not_perfectly_predicted <- all_dat %>%
  filter(treatment == "no_dms") %>% 
  filter(sen < 0.9 & ppv < 0.9) %>% 
  pull(id) %>% 
  unique()

plt_dat <- filter(all_dat, id %in% not_perfectly_predicted)
plt_dat <- filter(plt_dat, depth %in% c(15, 50))


d_dat <- plt_dat %>% filter(depth == 15, size > 20) %>% 
  select(-depth) %>% 
  filter(treatment == "dms")

ut_dat <- plt_dat %>% filter(depth == 15, size > 20) %>% 
  select(-depth) %>% 
  filter(treatment != "dms")

dat <- left_join(d_dat, ut_dat, 
                 by = c("size",
                        "score",
                        "id",
                        "seq",
                        "ref"),
                 suffix = c("_dms", "_ut"))

dat <- mutate(dat, 
              delta_ppv = ppv_dms - ppv_ut,
              delta_sen = sen_dms - sen_ut)

dat <- select(dat, 
              pred_dms,
              pred_ut, 
              ref, 
              everything())

dat <- filter(dat, 
              delta_ppv > 0.25, 
              delta_sen > 0.25)


library(txcompare)
gtf <- tidy_gtf("~/Projects/shared_dbases/annotation/human/gencode.v31lift37.basic.annotation.gtf") 
  
ref <- refs[[1]] %>% 
  group_by(strand)

gtf <- group_by(gtf, strand) 

ref_annotated <- bed_intersect(ref, gtf, suffix = c("", ".y"))

ref_annotated <- ref_annotated %>% 
  group_by(id) %>% 
  dplyr::slice(1) %>% 
  ungroup() %>% 
  mutate(id, new_id = str_c(gene_name.y, "_", chrom, ":", start, "-", end)) %>% 
  select(id, new_id) %>% 
  unique() %>% 
  group_by(id)

dat <- left_join(dat, ref_annotated, by = "id")
```

```{r}



plot_pred <- function(df, 
                      folding_dir = "all_new_cutoffs/15_rnastruct_mapfiles",
                      outdir = "."){
  map_dat <- read_map(file.path(folding_dir, 
                                paste0(df$id, "_coverage.map")))
  
  plot_secondary_structure(df$seq, 
                           df$ref,
                           color_by = map_dat$react,
                           outpath = file.path(outdir,
                                               paste0(df$new_id, 
                                                      "_evofold.png")),
                           windsorize = FALSE
                           )
  
  plot_secondary_structure(df$seq, 
                           df$pred_dms,
                           color_by = map_dat$react,
                           outpath = file.path(outdir,
                                               paste0(df$new_id, 
                                                      "_with_dms.png")),
                           windsorize = FALSE
                           )
  
    plot_secondary_structure(df$seq, 
                           df$pred_ut,
                           color_by = NULL,
                           outpath = file.path(outdir,
                                               paste0(df$new_id, 
                                                      "_without_dms.png")),
                           windsorize = FALSE
                           )
    
    outfilenames <- c(
      "_without_dms.png",
      "_with_dms.png",
      "_evofold.png")
    
    outfiles <-  file.path(outdir, paste0(df$new_id,  outfilenames))

  montage_args <- c(
    "-label",
    "%t",
    outfiles,
    "-pointsize", 128,
    "-tile", "x1",
    "-geometry", "+2+2",
    file.path(outdir, paste0(df$new_id, "_combined.png"))) 

  system2("montage",
        montage_args)
  
  unlink(outfiles)

}


to_plot <- c(
  #"2002388.10_1_-_200",
  "1098335_0_+_74",
  "302307.3_0_+_100",
  "1547267_0_+_79",
  "1047271_0_+_69",
  "1112227_1_+_64"
)

# to_plot <- c(
#   "1022231_0_+_95", # CLTC CDS
#   "1495619_0_+_59", # HMGXB4 3UTR
#   "491952.0_0_-_92", # FBXL14 CDS
#   "1932934_0_-_135" #FBN2 CDS)
# )
plt_dat <- filter(dat, id %in% to_plot)

outdir = "plots/rnastructure/all_nucs/examples"
dir.create(outdir)
walk(split(plt_dat, 1:nrow(plt_dat)),
     ~plot_pred(.x, outdir = outdir))
```

